internalsdk/auth/srp.go

```diff
@@ -0,0 +1,188 @@
+	return srp.NewSRPClient(srp.KnownGroups[group], encryptedKey, nil)
+}
+
+func ConvertToUserDetailsResponse(userResponse *protos.LoginResponse) *protos.User {
+	// Convert protobuf to usre details struct
+	log.Debugf("ConvertToUserDetailsResponse %+v", userResponse)
+
+	user := userResponse.LegacyUserData
+
+	userData := protos.User{
+		UserId:       userResponse.LegacyUserData.UserId,
+		Code:         user.Code,
+		Token:        userResponse.LegacyToken,
+		Referral:     user.Code,
+		UserLevel:    user.UserLevel,
+		Expiration:   user.Expiration,
+		Email:        user.Email,
+		UserStatus:   user.UserStatus,
+		Locale:       user.Locale,
+		YinbiEnabled: user.YinbiEnabled,
+		Inviters:     user.Inviters,
+		Invitees:     user.Invitees,
+		// Purchases:    user.Purchases,
+	}
+	log.Debugf("ConvertToUserDetailsResponse %+v", &userData)
+
+	for _, d := range user.Devices {
+		// Map the fields from LoginResponse_Device to UserDevice
+		userDevice := &protos.Device{
+			Id:      d.Id,
+			Name:    d.GetName(),
+			Created: d.GetCreated(),
+		}
+		userData.Devices = append(userData.Devices, userDevice)
+	}
+
+	return &userData
+}
+
+// Takes password and email, salt and returns encrypted key
+func GenerateEncryptedKey(password string, email string, salt []byte) *big.Int {
+	if len(salt) == 0 || len(password) == 0 || len(email) == 0 {
+		log.Errorf("slat or password or email is empty %v %v %v", salt, password, email)
+		return nil
+	}
+	lowerCaseEmail := strings.ToLower(email)
+	combinedInput := password + lowerCaseEmail
+	encryptedKey := pbkdf2.Key([]byte(combinedInput), salt, 4096, 32, sha256.New)
+	encryptedKeyBigInt := big.NewInt(0).SetBytes(encryptedKey)
+	return encryptedKeyBigInt
+}
+
+func (c *authClient) getUserSalt(email string) ([]byte, error) {
+	lowerCaseEmail := strings.ToLower(email)
+	log.Debugf("Salt not found calling api for %s", email)
+	salt, err := c.GetSalt(context.Background(), lowerCaseEmail)
+	if err != nil {
+		return nil, err
+	}
+	log.Debugf("Salt Response-> %v", salt.Salt)
+	return salt.Salt, nil
+}
+
+func GenerateSalt() ([]byte, error) {
+	salt := make([]byte, 16)
+	if n, err := rand.Read(salt); err != nil {
+		return nil, err
+	} else if n != 16 {
+		return nil, errors.New("failed to generate 16 byte salt")
+	}
+	return salt, nil
+}
+
+func (c *authClient) SignUp(email string, password string) ([]byte, error) {
+	lowerCaseEmail := strings.ToLower(email)
+	salt, err := GenerateSalt()
+	if err != nil {
+		return nil, err
+	}
+
+	srpClient := NewSRPClient(lowerCaseEmail, password, salt)
+	verifierKey, err := srpClient.Verifier()
+	if err != nil {
+		return nil, err
+	}
+	signUpRequestBody := &protos.SignupRequest{
+		Email:                 lowerCaseEmail,
+		Salt:                  salt,
+		Verifier:              verifierKey.Bytes(),
+		SkipEmailConfirmation: true,
+	}
+	log.Debugf("Sign up request email %v, salt %v verifier %v verifiter in bytes %v", lowerCaseEmail, salt, verifierKey, verifierKey.Bytes())
```

Comment:
I think once we have confirmation this is working well, we should avoid logging the salt, verifier, etc. here in case the user submits their logs to us

Reply:
correct, I am planning to remove logs.